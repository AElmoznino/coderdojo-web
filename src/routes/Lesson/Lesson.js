// @flow

import React from 'react'
import styled from 'styled-components'
import type { Match } from 'react-router-dom'
import gql from 'graphql-tag'
import { Query } from 'react-apollo'
import ReactMarkdown from 'react-markdown'
import ErrorMessage from 'components/ErrorMessage/ErrorMessage'
import GridColumn from 'components/Grid/GridColumn'
import GridSection from 'components/Grid/GridSection'
import H1 from 'components/Typography/H1'
import H4 from 'components/Typography/H4'
import LoadingIndicator from 'components/LoadingIndicator/LoadingIndicator'
import NextLesson from './components/NextLesson'

type LessonProps = {
  match: Match,
}

const GET_LESSON = gql`
  query GetLesson($lesson: String) {
    lesson(where: { lessonId: $lesson }) {
      difficultyLevel
      lessonBody
      lessonIntro
      lessonObjectives
      lessonShortFacts
      lessonTitle
      nextLesson {
        lessonId
        lessonTitle
      }
    }
  }
`

const TopWrap = styled.div`
  display: grid;
  grid-template-columns: 2fr 1fr;
  grid-column-gap: 20px;
`

const ObjectivesBox = styled.div`
  background-color: ${({ theme }) => theme.colors.downy};
  padding: ${({ theme }) => theme.sizes.medium};
`

export const Lesson = ({ match }: LessonProps) => (
  <Query query={GET_LESSON} variables={{ lesson: match.params.lessonId }}>
    {({ data: { lesson }, error, loading }) => {
      if (loading) return <LoadingIndicator />
      if (error) return <ErrorMessage />
      if (!lesson) return null

      return (
        <GridColumn>
          <GridSection>
            <TopWrap>
              <div>
                <H1>{lesson.lessonTitle}</H1>
                {lesson.lessonShortFacts && <p>{lesson.lessonShortFacts}</p>}
                <p>{lesson.lessonIntro}</p>
              </div>

              <ObjectivesBox>
                <H4>Du kommer l√§ra dig:</H4>
                <ReactMarkdown>{lesson.lessonObjectives}</ReactMarkdown>
              </ObjectivesBox>
            </TopWrap>

            <ReactMarkdown escapeHtml={false}>
              {lesson.lessonBody}
            </ReactMarkdown>
          </GridSection>

          <NextLesson
            difficultyLevel={lesson.difficultyLevel}
            nextLesson={lesson.nextLesson}
          />
        </GridColumn>
      )
    }}
  </Query>
)

export default Lesson
