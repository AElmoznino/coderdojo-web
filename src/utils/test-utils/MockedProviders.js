import * as React from 'react'
import { ApolloProvider } from 'react-apollo'
import { BrowserRouter } from 'react-router-dom'
import { ThemeProvider } from 'styled-components'
import { theme } from '../../theme'
import ApolloClient from 'apollo-boost'
import createMockedNetworkFetch from 'apollo-mocknetworkinterface'

jest.unmock('react-apollo')
jest.unmock('react-router-dom')

type MockedProvidersProps = {
  children: React.Node,
  loading?: boolean,
  response?: Object,
}

const MockedProviders = ({
  children,
  loading,
  response = {},
}: MockedProvidersProps) => {
  const createResponse = () => {
    return {
      loading: !!loading,
      data: {
        ...response,
      },
    }
  }

  const mockedNetworkFetch = createMockedNetworkFetch(createResponse, {
    latency: 0,
  })

  const client = new ApolloClient({
    uri: 'http://localhost:4000',
    fetch: mockedNetworkFetch,
  })

  return (
    <ApolloProvider client={client}>
      <ThemeProvider theme={theme}>
        <BrowserRouter>{children}</BrowserRouter>
      </ThemeProvider>
    </ApolloProvider>
  )
}

export default MockedProviders
