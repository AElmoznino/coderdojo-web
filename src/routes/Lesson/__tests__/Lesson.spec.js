import React from 'react'
import { mount } from 'enzyme'
import Lesson from '../Lesson'
import MockedProviders from 'utils/test-utils/MockedProviders'

jest.unmock('graphql-tag')
jest.unmock('react-apollo')

describe('components/Lesson', () => {
  let component
  let match
  let mockedResponse

  beforeEach(() => {
    match = {
      params: {
        lessonId: 'nybörjare-1',
      },
    }

    mockedResponse = {
      lesson: {
        __typename: 'Lesson',
        difficultyLevel: 'nybörjare',
        lessonTitle: 'HTML del 1',
        lessonShortFacts: null,
        lessonIntro:
          'HTML används för att strukturera upp hemsidor. Alla hemsidor är byggda med det.',
        lessonBody:
          '## Vad är HTML\nHTML står för Hypertext Markup Language. Det är instruktioner för webbläsaren hur sidan ska ritas upp i din webbläsare. \n\nDe flesta hemsidor, även denna, har rubriker, textstycken, bilder, den s.k. navigationen högst upp. HTML instruerar webbläsaren hur den ska rita upp din hemsida.\n\n##Hur skriver man HTML\nHTML består av olika så kallade "taggar" som är förbestämda. Rubriker skrivs som  osv beroende på storlek. \nOm du vill skriva en rubrik, så måste du tänka på att själva rubriken ska vara *innanför* taggarna, såhär: \nI rutan nedan kan du se exempel på hur HTML-kod skrivs (till vänster), och hur det då kommer se ut (till höger).\n\n<iframe src="https://codesandbox.io/embed/v3qpwx7063?fontsize=12&hidenavigation=1" style="width:100%; height:500px; overflow:hidden;" sandbox="allow-modals allow-forms allow-popups allow-scripts allow-same-origin"></iframe>',
        lessonObjectives:
          'Du kommer lära dig:\n\n1. vad HTML är och används till\n2. strukturera upp din hemsida med HTML',
        nextLesson: {
          __typename: 'NextLesson',
          lessonId: 'nybörjare-2',
          lessonTitle: 'HTML del 2',
        },
      },
    }

    component = mount(
      <MockedProviders response={mockedResponse}>
        <Lesson match={match} />
      </MockedProviders>,
    )
  })

  it('renders Lesson', done => {
    setImmediate(() => {
      component.update()

      expect(component).toMatchSnapshot()

      done()
    })
  })
})
