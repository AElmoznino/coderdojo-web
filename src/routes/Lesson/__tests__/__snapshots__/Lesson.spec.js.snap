// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Lesson renders Lesson 1`] = `
<styled.div>
  <div
    className="sc-bdVaJa cBlAJt"
  >
    <styled.div>
      <div
        className="sc-bwzfXH LKvyz"
      >
        <styled.div>
          <div
            className="sc-htoDjs jUiOwB"
          >
            <div>
              <styled.h1>
                <h1
                  className="sc-htpNat kpiFDp"
                >
                  HTML del 1
                </h1>
              </styled.h1>
              <p>
                HTML används för att strukturera upp hemsidor. Alla hemsidor är byggda med det.
              </p>
            </div>
            <styled.div>
              <div
                className="sc-dnqmqq dXnOWs"
              >
                <styled.h4>
                  <h4
                    className="sc-ifAKCX dhEZNh"
                  >
                    Du kommer lära dig:
                  </h4>
                </styled.h4>
                <ol>
                  <styled.li
                    key="0"
                  >
                    <li
                      className="sc-iwsKbI kqXDDu"
                    >
                      strukturera upp din hemsida med HTML
                    </li>
                  </styled.li>
                  <styled.li
                    key="1"
                  >
                    <li
                      className="sc-iwsKbI kqXDDu"
                    >
                      inspektera andra hemsidor för att se hur de är uppbyggda
                    </li>
                  </styled.li>
                </ol>
              </div>
            </styled.div>
          </div>
        </styled.div>
        Array [
          <styled.h1>
            <h1
              className="sc-htpNat kpiFDp"
            >
              Vad är HTML
            </h1>
          </styled.h1>,
          <p
            key="paragraph-1-1"
          >
            HTML står för Hypertext Markup Language. Det är instruktioner för webbläsaren hur sidan ska ritas upp i din webbläsare.
          </p>,
        ]
        Array [
          <styled.h1>
            <h1
              className="sc-htpNat kpiFDp"
            >
              Hur skriver man HTML
            </h1>
          </styled.h1>,
          <p
            key="paragraph-1-1"
          >
            HTML består av olika så kallade "taggar" som är förbestämda. En rubrik skrivs som 
            &lt;h1&gt;
            , 
            &lt;h2&gt;
             osv beroende på storlek.
          </p>,
          <p>
            Här kommer mer text
          </p>,
        ]
      </div>
    </styled.div>
    <styled.div>
      <div
        className="sc-bwzfXH LKvyz"
      >
        <styled.div>
          <div
            className="sc-gzVnrw liWgEh"
          >
            <strong>
              Nästa lektion: 
              HTML del 2
            </strong>
            <Styled(Link)
              to="/lesson/nybörjare-2"
            >
              <a
                className="sc-bZQynM LpvqH"
                href="/lesson/nybörjare-2"
                onClick={[Function]}
              >
                Fortsätt &gt;
              </a>
            </Styled(Link)>
          </div>
        </styled.div>
      </div>
    </styled.div>
  </div>
</styled.div>
`;
